File: src/lib/schemas.ts
import { z } from 'zod'

// Settings schema
export const settingsSchema = z.object({
  difficulty: z.enum(['easy', 'normal', 'hard']).default('normal'),
  maxPlayers: z.number().min(1).max(100).default(10),
  timeLimit: z.number().min(0).default(300),
  friendlyFire: z.boolean().default(false),
  respawnTime: z.number().min(0).default(5),
})

// Weapon schema
export const weaponSchema = z.object({
  name: z.string().min(1),
  damage: z.number().min(0).max(1000).default(50),
  fireRate: z.number().min(0).default(1),
  recoil: z.number().min(0).max(1).default(0.1),
  ammoCapacity: z.number().min(0).default(30),
  reloadTime: z.number().min(0).default(2),
})

// Character schema
export const characterSchema = z.object({
  name: z.string().min(1),
  health: z.number().min(1).default(100),
  speed: z.number().min(0).default(10),
  jumpHeight: z.number().min(0).default(5),
  armor: z.number().min(0).default(0),
})

// Map schema
export const mapSchema = z.object({
  name: z.string().min(1),
  size: z.object({
    width: z.number().min(1),
    height: z.number().min(1),
  }).default({ width: 100, height: 100 }),
  gravity: z.number().min(0).default(9.8),
  lighting: z.enum(['day', 'night', 'dusk']).default('day'),
})

// Script schema
export const scriptSchema = z.object({
  name: z.string().min(1),
  content: z.string(),
})

// Texture schema
export const textureSchema = z.object({
  name: z.string().min(1),
  url: z.string().url(),
})

// AI response schema for structured changes
export const aiResponseSchema = z.object({
  type: z.enum(['settings', 'weapons', 'characters', 'maps', 'scripts']),
  changes: z.array(z.object({
    field: z.string(),
    value: z.any(),
    description: z.string().optional(),
  })),
})

// Union of all schemas
export const modSchemas = {
  settings: settingsSchema,
  weapons: weaponSchema,
  characters: characterSchema,
  maps: mapSchema,
  scripts: scriptSchema,
  textures: textureSchema,
}

File: backend/app/schemas.py
from typing import Literal, Optional, List, Dict, Any

from pydantic import BaseModel, Field

class UserBase(BaseModel):
    email: str

class UserCreate(UserBase):
    password: str

class User(UserBase):
    id: int
    is_active: bool

    class Config:
        orm_mode = True

class UserLogin(BaseModel):
    email: str
    password: str

class Token(BaseModel):
    access_token: str
    token_type: str

class AIChange(BaseModel):
    field: str
    value: Any
    description: Optional[str] = None

class AIResponse(BaseModel):
    type: Literal["settings", "weapons", "characters", "maps", "scripts"]
    changes: List[AIChange]

class AIPromptRequest(BaseModel):
    prompt: str
    project_id: Optional[int] = None

class ProjectBase(BaseModel):
    name: str
    description: Optional[str] = None

class ProjectCreate(ProjectBase):
    pass

class ProjectUpdate(ProjectBase):
    pass

class Project(ProjectBase):
    id: int
    user_id: int

    class Config:
        orm_mode = True

# Mod structures
class SettingsModel(BaseModel):
    difficulty: Literal["easy", "normal", "hard"] = "normal"
    maxPlayers: int = Field(10, ge=1, le=128)
    timeLimit: int = Field(300, ge=0, le=3600)
    friendlyFire: bool = False
    respawnTime: int = Field(5, ge=0, le=300)

class WeaponModel(BaseModel):
    name: str
    damage: float = Field(50, ge=0, le=1000)
    fireRate: float = Field(1, ge=0)
    recoil: float = Field(0.1, ge=0, le=1)
    ammoCapacity: int = Field(30, ge=0)
    reloadTime: float = Field(2, ge=0)

class CharacterModel(BaseModel):
    name: str
    health: float = Field(100, ge=1, le=1000)
    speed: float = Field(10, ge=0, le=500)
    jumpHeight: float = Field(5, ge=0, le=100)
    armor: float = Field(0, ge=0, le=1000)

class MapModel(BaseModel):
    name: str
    size: Dict[str, float]
    gravity: float = Field(9.8, ge=0, le=100)
    lighting: Literal["day", "night", "dusk"] = "day"

class ScriptModel(BaseModel):
    name: str
    content: str

class ModData(BaseModel):
    settings: Optional[SettingsModel] = None
    weapons: Optional[List[WeaponModel]] = None
    characters: Optional[List[CharacterModel]] = None
    maps: Optional[List[MapModel]] = None
    scripts: Optional[List[ScriptModel]] = None

class ModDataResponse(ModData):
    project_id: int

